distsql/distsql_test.go

```diff
@@ -14,32 +14,182 @@
+
+	response.Fetch(context.TODO())
+
+	// Test NextChunk.
+	chk := chunk.NewChunk(colTypes)
+	numAllRows := 0
+	for {
+		err = response.NextChunk(context.TODO(), chk)
+		c.Assert(err, IsNil)
+		numAllRows += chk.NumRows()
+		if chk.NumRows() == 0 {
+			break
+		}
+	}
+	c.Assert(numAllRows, Equals, 2)
+	err = response.Close()
+	c.Assert(err, IsNil)
 }
 
-func (resp *mockResponse) Next(ctx context.Context) ([]byte, error) {
-	resp.count++
-	if resp.count == 100 {
-		return nil, errors.New("error happened")
+func (s *testSuite) TestSelectStreaming(c *C) {
+	request, err := (&RequestBuilder{}).SetKeyRanges(nil).
+		SetDAGRequest(&tipb.DAGRequest{}).
+		SetDesc(false).
+		SetKeepOrder(false).
+		SetPriority(kv.PriorityNormal).
+		SetFromSessionVars(variable.NewSessionVars()).
+		SetStreaming(true).
+		Build()
+	c.Assert(err, IsNil)
+
+	/// 4 int64 types.
+	colTypes := []*types.FieldType{
+		{
+			Tp:      mysql.TypeLonglong,
+			Flen:    mysql.MaxIntWidth,
+			Decimal: 0,
+			Flag:    mysql.BinaryFlag,
+			Charset: charset.CharsetBin,
+			Collate: charset.CollationBin,
+		},
 	}
-	return mockSubresult(), nil
+	colTypes = append(colTypes, colTypes[0])
+	colTypes = append(colTypes, colTypes[0])
+	colTypes = append(colTypes, colTypes[0])
+
+	s.sctx.GetSessionVars().EnableStreaming = true
+
+	// Test NextChunk.
+	response, err := Select(context.TODO(), s.sctx, request, colTypes)
+	c.Assert(err, IsNil)
+	result, ok := response.(*streamResult)
+	c.Assert(ok, IsTrue)
+	c.Assert(result.rowLen, Equals, len(colTypes))
+
+	response.Fetch(context.TODO())
+
+	// Test NextChunk.
+	chk := chunk.NewChunk(colTypes)
+	numAllRows := 0
+	for {
+		err = response.NextChunk(context.TODO(), chk)
+		c.Assert(err, IsNil)
+		numAllRows += chk.NumRows()
+		if chk.NumRows() == 0 {
+			break
+		}
+	}
+	c.Assert(numAllRows, Equals, 2)
+	err = response.Close()
+	c.Assert(err, IsNil)
+}
+
+func (s *testSuite) TestAnalyze(c *C) {
+	request, err := (&RequestBuilder{}).SetKeyRanges(nil).
+		SetAnalyzeRequest(&tipb.AnalyzeReq{}).
+		SetKeepOrder(true).
+		SetPriority(kv.PriorityLow).
+		Build()
+	c.Assert(err, IsNil)
+
+	response, err := Analyze(context.TODO(), s.sctx.GetClient(), request)
+	c.Assert(err, IsNil)
+
+	result, ok := response.(*selectResult)
+	c.Assert(ok, IsTrue)
+	c.Assert(result.label, Equals, "analyze")
+
+	response.Fetch(context.TODO())
+
+	bytes, err := response.NextRaw(context.TODO())
+	c.Assert(err, IsNil)
+	c.Assert(len(bytes), Equals, 14)
+
+	err = response.Close()
+	c.Assert(err, IsNil)
 }
 
+// mockResponse implements kv.Response interface.
+// Used only for test.
+type mockResponse struct{ count int }
+
+// Close implements kv.Response interface.
 func (resp *mockResponse) Close() error {
+	resp.count = 0
 	return nil
 }
 
-func mockSubresult() []byte {
-	resp := new(tipb.SelectResponse)
-	b, err := resp.Marshal()
+// Next implements kv.Response interface.
+func (resp *mockResponse) Next(ctx context.Context) (kv.ResultSubset, error) {
+	if resp.count == 2 {
```

Comment:
This snippet confuses me. Why are u directly returning nil value to its caller when `count` is 2? may be you need comment it?

Reply:
It's mocked, we only return 2 rows for convenience. However you can return as many rows as you wish.